firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java

```diff
@@ -359,17 +368,30 @@ private void ensureSequenceNumbers() {
     SQLiteStatement tagDocument =
         db.compileStatement(
             "INSERT INTO target_documents (target_id, path, sequence_number) VALUES (0, ?, ?)");
+
     SQLitePersistence.Query untaggedDocumentsQuery =
         new SQLitePersistence.Query(
-            db,
-            "SELECT RD.path FROM remote_documents AS RD WHERE NOT EXISTS (SELECT TD.path FROM target_documents AS TD WHERE RD.path = TD.path AND TD.target_id = 0)");
-    untaggedDocumentsQuery.forEach(
-        row -> {
-          tagDocument.clearBindings();
-          tagDocument.bindString(1, row.getString(0));
-          tagDocument.bindLong(2, sequenceNumber);
-          hardAssert(tagDocument.executeInsert() != -1, "Failed to insert a sentinel row");
-        });
+                db,
+                "SELECT RD.path FROM remote_documents AS RD WHERE NOT EXISTS ("
+                    + "SELECT TD.path FROM target_documents AS TD "
+                    + "WHERE RD.path = TD.path AND TD.target_id = 0"
+                    + ") LIMIT ?")
+            .binding(SEQUENCE_NUMBER_BATCH_SIZE);
+
+    boolean[] resultsRemaining = new boolean[1];
+
+    do {
+      resultsRemaining[0] = false;
+
+      untaggedDocumentsQuery.forEach(
+          row -> {
+            resultsRemaining[0] = true;
+            tagDocument.clearBindings();
+            tagDocument.bindString(1, row.getString(0));
+            tagDocument.bindLong(2, sequenceNumber);
+            hardAssert(tagDocument.executeInsert() != -1, "Failed to insert a sentinel row");
+          });
+    } while (resultsRemaining[0]);
```

Comment:
This will always do at least one extra query.  You could avoid that by counting the rows and comparing to SEQUENCE_NUMBER_BATCH_SIZE.  But my guess is you know this and chose to keep it simpler, which is also fine.

Reply:
I don't think it is worth the extra code to handle this case, since this will only ever run once.